[Backtest] OrderMatchingEngine bar execution generates fractional fill quantities ignoring instrument size_precision

# Bug Report

## Confirmation

- [x] I've re-read the relevant sections of the documentation.
- [x] I've searched existing issues and discussions to avoid duplicates.
- [x] I've reviewed or skimmed the source code (or examples) to confirm the behavior is not by design.
- [ ] I've tested this issue using a recent *development* wheel (`dev` develop or `a` nightly) and can still reproduce it.

> Note: Dev wheels are not available for Python 3.12 (my environment). Tested on v1.222.0 (latest stable).

### Expected Behavior

When running bar-based backtests with `bar_execution=True`, fill quantities should be rounded to match the instrument's `size_precision`. For an instrument with `size_precision=0` (integer quantities only), fills should be whole numbers.

### Actual Behavior

The matching engine generates fill quantities derived directly from `bar.volume / 4` without rounding to instrument precision. This produces fractional fill quantities for instruments that only allow integers.

**Example:**
- Instrument: US Equity with `size_precision=0`, `size_increment=1`
- Bar volume: 150
- Available volume per OHLC price level: `150 / 4 = 37.5`
- Order: BUY 51 units (IOC)
- Fill: **37.5 units** ❌ (should be 37 or 38)

This causes cascading issues:
1. `order.leaves_qty` becomes fractional (13.5)
2. Position `signed_qty` becomes fractional
3. Eventually triggers `AssertionError: assert side != PositionSide.FLAT` in `PositionOpened.__init__` when position state becomes corrupted

### Steps to Reproduce the Problem

1. Create a backtest with bar data for an instrument with `size_precision=0`
2. Use default venue config (`bar_execution=True`, `book_type=L1_MBP`)
3. Submit a market order with `TimeInForce.IOC` where quantity > `bar.volume / 4`
4. Observe fractional fill quantity in `OrderFilled` event

### Code Snippets or Logs

```python
from nautilus_trader.backtest.engine import BacktestEngine, BacktestEngineConfig
from nautilus_trader.model.enums import OrderSide, TimeInForce

# Setup engine with venue (bar_execution=True by default)
# Load instrument with size_precision=0
# Load 1-minute bars

# In strategy on_bar():
order = self.order_factory.market(
    instrument_id=instrument.id,
    order_side=OrderSide.BUY,
    quantity=instrument.make_qty(51),  # Integer quantity
    time_in_force=TimeInForce.IOC,
)
self.submit_order(order)

# OrderFilled event shows:
#   last_qty: 37.5  <-- FRACTIONAL despite size_precision=0
#   leaves_qty: 13.5
```

Debug output from investigation:
```
*** on_order_filled ***
  order.quantity: 51
  order.filled_qty: 37.5   <-- Fractional!
  order.leaves_qty: 13.5   <-- Fractional!
  RAW: quantity.raw=510000000000000000, filled.raw=375000000000000000, leaves.raw=135000000000000000
```

### Suggested Fix

In the matching engine, when calculating fill quantity from bar volume, round to the instrument's size increment:

```python
# Current (simplified):
fill_qty = min(order.leaves_qty, available_volume)

# Suggested:
fill_qty = instrument.make_qty(min(order.leaves_qty, available_volume))
```

### Related

- #2275 fixed sizes *below* minimum increment (added `max(..., 1)`)
- This issue: sizes *above* minimum but fractional are not rounded to `size_increment`

### Specifications

- OS platform: macOS Darwin 22.3.0 (arm64)
- Python version: 3.12.11
- `nautilus_trader` version: 1.222.0

--- a/nautilus_trader/backtest/engine.pyx
+++ b/nautilus_trader/backtest/engine.pyx
@@ -4030,10 +4030,16 @@ cdef class OrderMatchingEngine:
             quarter_raw = min_size_raw
 
         # Calculate close size: remaining volume after 3 quarters, also rounded
-        cdef QuantityRaw close_raw = bar._mem.volume.raw - (quarter_raw * 3)
-        close_raw = (close_raw // min_size_raw) * min_size_raw
-        if close_raw < min_size_raw:
-            close_raw = min_size_raw
+        # Protect against underflow when quarter_raw * 3 exceeds bar volume
+        cdef QuantityRaw three_quarters = quarter_raw * 3
+        cdef QuantityRaw close_raw
+        if three_quarters >= bar._mem.volume.raw:
+            close_raw = min_size_raw  # Use minimum if no remaining volume
+        else:
+            close_raw = bar._mem.volume.raw - three_quarters
+            close_raw = (close_raw // min_size_raw) * min_size_raw
+            if close_raw < min_size_raw:
+                close_raw = min_size_raw
 
         cdef Quantity size = Quantity.from_raw_c(quarter_raw, bar._mem.volume.precision)
         cdef Quantity close_size = Quantity.from_raw_c(close_raw, bar._mem.volume.precision)
@@ -4143,14 +4149,26 @@ cdef class OrderMatchingEngine:
             ask_quarter = min_size_raw
 
         # Calculate close sizes: remaining volume after 3 quarters, also rounded
-        cdef QuantityRaw bid_close_raw = self._last_bid_bar._mem.volume.raw - (bid_quarter * 3)
-        bid_close_raw = (bid_close_raw // min_size_raw) * min_size_raw
-        if bid_close_raw < min_size_raw:
-            bid_close_raw = min_size_raw
-        cdef QuantityRaw ask_close_raw = self._last_ask_bar._mem.volume.raw - (ask_quarter * 3)
-        ask_close_raw = (ask_close_raw // min_size_raw) * min_size_raw
-        if ask_close_raw < min_size_raw:
-            ask_close_raw = min_size_raw
+        # Protect against underflow when quarter * 3 exceeds bar volume
+        cdef QuantityRaw bid_three_quarters = bid_quarter * 3
+        cdef QuantityRaw bid_close_raw
+        if bid_three_quarters >= self._last_bid_bar._mem.volume.raw:
+            bid_close_raw = min_size_raw  # Use minimum if no remaining volume
+        else:
+            bid_close_raw = self._last_bid_bar._mem.volume.raw - bid_three_quarters
+            bid_close_raw = (bid_close_raw // min_size_raw) * min_size_raw
+            if bid_close_raw < min_size_raw:
+                bid_close_raw = min_size_raw
+
+        cdef QuantityRaw ask_three_quarters = ask_quarter * 3
+        cdef QuantityRaw ask_close_raw
+        if ask_three_quarters >= self._last_ask_bar._mem.volume.raw:
+            ask_close_raw = min_size_raw  # Use minimum if no remaining volume
+        else:
+            ask_close_raw = self._last_ask_bar._mem.volume.raw - ask_three_quarters
+            ask_close_raw = (ask_close_raw // min_size_raw) * min_size_raw
+            if ask_close_raw < min_size_raw:
+                ask_close_raw = min_size_raw
 
         cdef Quantity bid_size = Quantity.from_raw_c(bid_quarter, self._last_bid_bar._mem.volume.precision)
         cdef Quantity ask_size = Quantity.from_raw_c(ask_quarter, self._last_ask_bar._mem.volume.precision)

